extensions/amp-story-panning-media/0.1/amp-story-panning-media.js

```diff
@@ -156,17 +183,74 @@ export class AmpStoryPanningMedia extends AMP.BaseElement {
   }
 
   /** @private */
-  onPageNavigation_() {
-    if (this.isOnActivePage_) {
-      this.animate_();
+  setAnimateTo_() {
+    const x = parseFloat(this.element_.getAttribute('x') || 0);
+    const y = parseFloat(this.element_.getAttribute('y') || 0);
+    const zoom = parseFloat(this.element_.getAttribute('zoom') || 1);
+    const lockBounds = this.element_.hasAttribute('lock-bounds');
+
+    if (!lockBounds) {
+      this.animateTo_ = {x, y, zoom};
+    } else {
+      // Zoom must be set to calculate maxBounds.
+      this.animateTo_.zoom = zoom < 1 ? 1 : zoom;
+
+      this.setMaxBounds_();
+
+      this.animateTo_.x =
+        x > 0
+          ? Math.min(this.maxBounds_.left, x)
+          : Math.max(this.maxBounds_.right, x);
+
+      this.animateTo_.y =
+        y > 0
+          ? Math.min(this.maxBounds_.top, y)
+          : Math.max(-this.maxBounds_.bottom, y);
     }
   }
 
+  /**
+   * Calculates max distances to keep image in viewport.
+   * This is only set if lock-bounds, left, right, top or bottom are specified.
+   * @private
+   */
+  setMaxBounds_() {
+    // Calculations to clamp image to edge of container.
+    const {width: containerWidth, height: containerHeight} = this.pageSize_;
+
+    const ampImgWidth = this.ampImgEl_.getAttribute('width');
```

Comment:
Could there be amp-img used without `width` / `height` explicit attributes? I don't think it's enforced by validation

Reply:
We need it for `getMaxBounds_`. I added a `user().error` if width and height aren't specified and `lock-bounds`.
I'll also look into how this can be validated.